baseplate/server/queue_consumer.py

```diff
@@ -0,0 +1,268 @@
+    """Factory for building all of the objects needed to run a QueueConsumerServer.
+
+    This is the type of object that you build in you `service.make_queue_consumer`
+    function to be passed to `make_server`.
+    """
+
+    @abc.abstractmethod
+    def build_pump_worker(self, work_queue: queue.Queue) -> PumpWorker:
+        """Build an object implementing the PumpWorker interface.
+
+        `work_queue` is the Queue that will be shared between the PumpWorker and
+        and the QueueConsumer to pass messages to the MessageHandler, be sure
+        that your PumpWorker uses `work_queue`.
+        """
+
+    @abc.abstractmethod
+    def build_message_handler(self) -> MessageHandler:
+        """Build an object implementing the MessageHandler interface."""
+
+    @abc.abstractmethod
+    def build_health_checker(self, listener: socket.socket) -> StreamServer:
+        """Build an HTTP server to service health checks."""
+
+
+class QueueConsumer:
+    """Wrapper around a MessageHandler object that interfaces with the work_queue and starts/stops the handle loop.
+
+    This object is used by the QueueConsumerServer to wrap a MessageHandler object
+    before creating a worker Thread.  This allows the MessageHandler to focus soley
+    on handling a single message while QueueConsumer pull messages from the work_queue
+    and handles starting/stopping based on the commands from the server.
+    """
+
+    def __init__(self, work_queue: queue.Queue, message_handler: MessageHandler):
+        self.id = uuid.uuid4()
+        self.work_queue = work_queue
+        self.message_handler = message_handler
+        self.started = False
+        self.stopped = False
+        self._queue_timeout = 5
+
+    def stop(self) -> None:
+        """Signal the QueueConsumer to stop processing."""
+        assert self.started
+        assert not self.stopped
+        self.stopped = True
+
+    def run(self) -> None:
+        """Run the queue consumer until stopped or hit an unhandled Exception."""
+        assert not self.started
+        assert not self.stopped
+        logger.debug("Consumer <%s> starting.", self.id)
+        self.started = True
+        while not self.stopped:
+            try:
+                # We set a timeout so we can periodically check if we should
+                # stop, this way we will actually return if we have recieved a
+                # `stop()` call rather than hanging in a `get` waiting for a
+                # new message.  If we did not do this, we would not be able to
+                # wait for all of our workers to finish before stopping the
+                # server.
+                message = self.work_queue.get(timeout=self._queue_timeout)
+            except queue.Empty:
+                pass
+            else:
+                # Ensure that if self.message_handler.handle throws a `queue.Empty`
+                # error, that bubbles up and is not treated as though `self.work_queue`
+                # is empty
```

Comment:
is this because we don't know anything about how `message_handler.handle` works so it could be running its own queue and hit `queue.Empty`?

Reply:
Correct.